Feat/add hpu support

PiotrBLL opened the pull request:
This pull request allows users to run the Whisper code on HPUs (Intel® Gaudi®):

- Created Dockerfile_hpu for HPU operations
- Updated the transcribe.py script to accommodate HPU-related changes
- Added HPU support to the README.md

PiotrBLL committed:
Add initial imports and usage of `wrap_in_hpu_graph`

PiotrBLL committed:
Add requires_hpu marker in tests

PiotrBLL committed:
Add testing hpu file

PiotrBLL committed:
Update requirements

PiotrBLL committed:
Add HPU support in transcribe and timing + tests

PiotrBLL committed:
Add hpu_utils

orionsBeltWest reviewed (changes requested):
Could you please add instruction for building Gaudi images or install steps. 

Probably an hpu Dockerfile or simple install steps?

sjagtap1803 commented on whisper/__init__.py:
```diff
@@ -6,6 +6,7 @@
 from typing import List, Optional, Union
 
 import torch
+from habana_frameworks.torch.utils.library_loader import load_habana_module
```
Does this work on machines/containers without habana_frameworks package? Wondering if load_habana_module and wrap_in_hpu_graph can be imported only if `device=="hpu"`.

PiotrBLL replied:
Thanks, thats true, changed to be loaded on condition - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-ccf3e38e2062ec2b77cbeda0fb709b927859e107a1d84463651fd53915182616R47

sjagtap1803 commented on whisper/hpu_utils.py:
```diff
@@ -0,0 +1,15 @@
+    Load HPU if available, otherwise use CUDA or CPU.
+    """
+
+    if not torch.hpu.is_available():
```
Does `torch.hpu.is_available()` work (returns False) if habana module hasn't been loaded (`load_habana_module`)?

PiotrBLL replied:
chaned it to load the `load_habana_module` at the very begging of the file and load that file only for specific use case with running `hpu`. E.g. test of new dtw hpu -> https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-50c05a2d91bd5551fadde4ec1249407f878e14f5ef9e52df34139d491f98f952R104

orionsBeltWest replied:
Now  the container build fine. 

But when I launch it as per readme.md 

docker run -it --runtime=habana \
  -e HABANA_VISIBLE_DEVICES=all \
  -e OMPI_MCA_btl_vader_single_copy_mechanism=none \
  --cap-add=sys_nice \
  --net=host \
  --ipc=host \
  -v /path/to/your/whisper:/workspace/whisper \
  whisper_hpu:latest \
  /bin/bash
  
  
  Then try:
  
  whisper audio.flac audio.mp3 audio.wav --model turbo --device hpu 
  
  It does not work! 
  
  ----------------
  
  Issues I notice:
  
  The Dockerfile set work dir to /root/whisper 
  
  but docker launcher map local whisper to /workspace/whisper 
  
  Is this okay?

PiotrBLL replied:
@orionsBeltWest Updated the README.md and Dockerfile.hpu

Ramji-Satya commented on tests/test_transcribe.py:
```diff
@@ -10,7 +10,7 @@
 @pytest.mark.parametrize("model_name", whisper.available_models())
 def test_transcribe(model_name: str):
     device = "cuda" if torch.cuda.is_available() else "cpu"
```
we should add a simlar test for hpu as well

PiotrBLL replied:
Thanks, just added a new test for `hpu` - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-ccf3e38e2062ec2b77cbeda0fb709b927859e107a1d84463651fd53915182616R47

orionsBeltWest reviewed (changes requested):
reinstalling torch with this requirement could override the gaudi touch?

orionsBeltWest commented on requirements.txt:
```diff
@@ -5,3 +5,5 @@ tqdm
 more-itertools
 tiktoken
 triton>=2.0.0;platform_machine=="x86_64" and sys_platform=="linux" or sys_platform=="linux2"
```
reinstalling torch could override the Gaudi torch already installed?

PiotrBLL replied:
I think it shouldn't override

orionsBeltWest replied:
I think torch will be reinstalled: 
could you show your install and launch steps?

PiotrBLL replied:
@orionsBeltWest 
I just added the [Dockerfile_hpu](https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-28de834099a6f1c52d83e6dff545f6c282de474af492ae17f8875437373c10d9)  
And info about [the new section in README.md](https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)  

In the Dockerfile_hpu I added installing packages just with adding:  
```txt
# Install Python packages from Whisper requirements
RUN pip install --upgrade pip \
    && pip install -r requirements.txt \
```
and the requirements.txt are:  
```txt
numba
numpy
torch
tqdm
more-itertools
tiktoken
triton>=2.0.0;platform_machine=="x86_64" and sys_platform=="linux" or sys_platform=="linux2"
scipy
pytest
```
I'm not sure if in this case, PyTorch will be overwritten, which could potentially cause issues with HPU functionality on the machine. Therefore, we also have the idea to avoid installing packages from `requirements.txt` and instead add the following lines at the end of the `Dockerfile_hpu`:  
```txt
# Install Python packages from Whisper requirements
RUN pip install --upgrade pip \
    && pip install numba==0.60.0 \
    && pip install numpy==2.1.3 \
    && pip install tqdm==4.66.6 \
    && pip install more-itertools==10.5.0 \
    && pip install tiktoken==0.8.0 \
    && pip install scipy==1.14.1 \
    && pip install optimum[habana]==1.13.1 \
```

orionsBeltWest replied:
Take a look at vllm repo It has a good example of specifying various requirements.txt files and Dockerfile for various accelerators. https://github.com/vllm-project/vllm

PiotrBLL replied:
Thank you for providing this repository. I have also adjusted the current Dockerfile to integrate with Whisper, including the installation of `ffmpeg` - is is mentioned in README.md by Whisper. It seems that the current version is now correct. What do you think about it?

PiotrBLL committed:
Fix importing habana-frameworks library conditionally

PiotrBLL committed:
Add hpu transcribe test

orionsBeltWest commented:
> Could you please add instruction for building Gaudi images or install steps.
> 
> Probably an hpu Dockerfile or simple install steps?

How do you run this repo with Gaudi? Could you provide the steps?

PiotrBLL committed:
Add Dockerfile_hpu + README.md section of hpu usage

PiotrBLL committed:
Add hpu usage info in jupyter notebook files

PiotrBLL committed:
Add Pytorch Docker images info in README.md

orionsBeltWest commented on Dockerfile_hpu:
```diff
@@ -0,0 +1,35 @@
+
+# Install Python packages from Whisper requirements
+RUN pip install --upgrade pip \
+    && pip install -r requirements.txt \
```
This Dockerfile is not building! 

Remove the \ from line 35 
The build is failing for not finding requiremets.txt Do you need to copy this? 
 Do we need git on line 18? 

Finally recommend renaming Dockerfile_hpu to Dockerfile.hpu.

PiotrBLL replied:
@orionsBeltWest 
Hey, sorry for the oversight and problems. Update:
- I fixed the Dockerfile
- Deleted installing git from line 18
- I changed the name to Dockerfile.hpu
- I updated the README.md

PiotrBLL committed:
Rename and Fix Dockerfile.hpu, add .dockerignore, update in README.md

PiotrBLL committed:
Add optimum[habana]==1.13.1 in Dockerfile.hpu

PiotrBLL committed:
Fix Dockerfile.hpu `ffmpeg` package install, update README.md whisper container usage

orionsBeltWest commented:
Built a new image and launched it. 

Ran the cmd from the readme:
python3 -m whisper.transcribe audio.flac audio.mp3 audio.wav --model turbo --device hpu

Partial crash dump below:


  libpostproc    58.  1.100 / 58.  1.100
[in#0 @ 0x8669d00] Error opening input: No such file or directory
Error opening input file audio.flac.
Error opening input files: No such file or directory

Skipping audio.flac due to RuntimeError: Failed to load audio: ffmpeg version 7.0.2-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gmp --enable-libgme --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libdav1d --enable-libxvid --enable-libzvbi --enable-libzimg
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
[in#0 @ 0x8669d00] Error opening input: No such file or directory
Error opening input file audio.flac.
Error opening input files: No such file or directory

/root/whisper/transcribe.py:126: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")
Traceback (most recent call last):
  File "/root/whisper/audio.py", line 58, in load_audio
    out = run(cmd, capture_output=True, check=True).stdout
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ffmpeg', '-nostdin', '-threads', '0', '-i', 'audio.mp3', '-f', 's16le', '-ac', '1', '-acodec', 'pcm_s16le', '-ar', '16000', '-']' returned non-zero exit status 254.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/whisper/transcribe.py", line 613, in cli
    result = transcribe(model, audio_path, temperature=temperature, **args)
  File "/root/whisper/transcribe.py", line 135, in transcribe
    mel = log_mel_spectrogram(audio, model.dims.n_mels, padding=N_SAMPLES)
  File "/root/whisper/audio.py", line 140, in log_mel_spectrogram
    audio = load_audio(audio)
  File "/root/whisper/audio.py", line 60, in load_audio
    raise RuntimeError(f"Failed to load audio: {e.stderr.decode()}") from e
RuntimeError: Failed to load audio: ffmpeg version 7.0.2-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gmp --enable-libgme --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libdav1d --enable-libxvid --enable-libzvbi --enable-libzimg
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
[in#0 @ 0x7f6cd00] Error opening input: No such file or directory
Error opening input file audio.mp3.
Error opening input files: No such file or directory

Skipping audio.mp3 due to RuntimeError: Failed to load audio: ffmpeg version 7.0.2-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers

PiotrBLL committed:
Fix hpu device_name typo

orionsBeltWest commented:
No getting new errors:

 PT_HPU_LAZY_MODE = 1
 PT_RECIPE_CACHE_PATH = 
 PT_CACHE_FOLDER_DELETE = 0
 PT_HPU_RECIPE_CACHE_CONFIG = 
 PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
 PT_HPU_LAZY_ACC_PAR_MODE = 0
 PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
---------------------------: System Configuration :---------------------------
Num CPU Cores : 152
CPU RAM       : 1056439512 KB
------------------------------------------------------------------------------
/root/whisper/transcribe.py:126: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")
Traceback (most recent call last):
  File "/root/whisper/audio.py", line 58, in load_audio
    out = run(cmd, capture_output=True, check=True).stdout
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ffmpeg', '-nostdin', '-threads', '0', '-i', 'audio.flac', '-f', 's16le', '-ac', '1', '-acodec', 'pcm_s16le', '-ar', '16000', '-']' returned non-zero exit status 254.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/whisper/transcribe.py", line 613, in cli
    result = transcribe(model, audio_path, temperature=temperature, **args)
  File "/root/whisper/transcribe.py", line 135, in transcribe
    mel = log_mel_spectrogram(audio, model.dims.n_mels, padding=N_SAMPLES)
  File "/root/whisper/audio.py", line 140, in log_mel_spectrogram
    audio = load_audio(audio)
  File "/root/whisper/audio.py", line 60, in load_audio
    raise RuntimeError(f"Failed to load audio: {e.stderr.decode()}") from e
RuntimeError: Failed to load audio: ffmpeg version 7.0.2-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gmp --enable-libgme --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libdav1d --enable-libxvid --enable-libzvbi --enable-libzimg
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
[in#0 @ 0x7723d00] Error opening input: No such file or directory
Error opening input file audio.flac.
Error opening input files: No such file or directory

PiotrBLL committed:
Fix warnings

PiotrBLL committed:
Wrap model to HPU + update Dockerfile packages

PiotrBLL committed:
Upgrade the Dockerfile.hpu ubuntu image

PiotrBLL committed:
Add checking sparse tensors in Whisper model

PiotrBLL committed:
Add util checking hpu tensor

PiotrBLL committed:
Add HPU checking in decoding and transcribe handle

PiotrBLL committed:
Fix: compute_device name in Whisper model

orionsBeltWest commented:
Here are some comments:

I do not see pip install command for whisper in the docker file?

after 
ADD . /root/whisper
I was expecting something like:

pip install . 

In the README launch command, you are mounting the local whisper dir to /root/whisper which you just copied!!  

Finally mounting the local dir is casing issue when rebuilding the image:

docker build --no-cache  -t sapdai/refd:whisper_gaudi-1.17.0 -f Dockerfile.hpu .
[+] Building 0.2s (6/11)                                                                                                                                                                       docker:default
 => [internal] load build definition from Dockerfile.hpu                                                                                                                                                 0.0s
 => => transferring dockerfile: 1.62kB                                                                                                                                                                   0.0s
 => [internal] load metadata for vault.habana.ai/gaudi-docker/1.18.0/ubuntu22.04/habanalabs/pytorch-installer-2.4.0:latest                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                        0.0s
 => => transferring context: 53B                                                                                                                                                                         0.0s
 => CACHED [1/7] FROM vault.habana.ai/gaudi-docker/1.18.0/ubuntu22.04/habanalabs/pytorch-installer-2.4.0:latest                                                                                          0.0s
 => ERROR [internal] load build context                                                                                                                                                                  0.0s
 => => transferring context: 69.98kB                                                                                                                                                                     0.0s
 => CANCELED [2/7] RUN apt-get update && apt-get install -y     tzdata     bash-completion     python3-pip     openssh-server     vim     iputils-ping     net-tools     protobuf-compiler     curl      0.2s
------
 > [internal] load build context:
------
ERROR: failed to solve: error from sender: open whisper/.graph_dumps: permission denied

I had to remove whisper/.graph_dumps to build again!



also please create new requirements_hpu.txt and remove torch from it! 
Is the optimum-habana already in the container

FROM vault.habana.ai/gaudi-docker/1.18.0/ubuntu22.04/habanalabs/pytorch-installer-2.4.0:latest

orionsBeltWest commented:
Finally here is the issue with your latest update:

root@0f213d1cffaf:~/whisper# python3 -m whisper.transcribe audio.flac audio.mp3 audio.wav --model turbo --device hpu
/usr/lib/python3.10/runpy.py:126: RuntimeWarning: 'whisper.transcribe' found in sys.modules after import of package 'whisper', but prior to execution of 'whisper.transcribe'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
/usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
  return isinstance(object, types.FunctionType)
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
100%|█████████████████████████████████████| 1.51G/1.51G [00:19<00:00, 83.2MiB/s]
/root/whisper/__init__.py:154: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(fp, map_location="cpu")
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
============================= HABANA PT BRIDGE CONFIGURATION =========================== 
 PT_HPU_LAZY_MODE = 1
 PT_RECIPE_CACHE_PATH = 
 PT_CACHE_FOLDER_DELETE = 0
 PT_HPU_RECIPE_CACHE_CONFIG = 
 PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
 PT_HPU_LAZY_ACC_PAR_MODE = 0
 PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
 PT_HPU_EAGER_PIPELINE_ENABLE = 1
 PT_HPU_EAGER_COLLECTIVE_PIPELINE_ENABLE = 1
---------------------------: System Configuration :---------------------------
Num CPU Cores : 152
CPU RAM       : 1056439512 KB
------------------------------------------------------------------------------
/root/whisper/transcribe.py:132: UserWarning: FP16 is not supported on HPU; using FP32 instead
  warnings.warn("FP16 is not supported on HPU; using FP32 instead")
Traceback (most recent call last):
  File "/root/whisper/audio.py", line 58, in load_audio
    out = run(cmd, capture_output=True, check=True).stdout
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ffmpeg', '-nostdin', '-threads', '0', '-i', 'audio.flac', '-f', 's16le', '-ac', '1', '-acodec', 'pcm_s16le', '-ar', '16000', '-']' returned non-zero exit status 254.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/whisper/transcribe.py", line 617, in cli
    result = transcribe(model, audio_path, temperature=temperature, **args)
  File "/root/whisper/transcribe.py", line 139, in transcribe
    mel = log_mel_spectrogram(audio, model.dims.n_mels, padding=N_SAMPLES)
  File "/root/whisper/audio.py", line 140, in log_mel_spectrogram
    audio = load_audio(audio)
  File "/root/whisper/audio.py", line 60, in load_audio
    raise RuntimeError(f"Failed to load audio: {e.stderr.decode()}") from e
RuntimeError: Failed to load audio: ffmpeg version 7.0.2-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)

PiotrBLL committed:
Add requirements_hpu.txt

PiotrBLL commented:
@orionsBeltWest 
We have added a new `requirements_hpu.txt` with key packages:

```plaintext
optimum-habana==1.14.1
transformers==4.45.2
huggingface-hub==0.26.2
tiktoken==0.8.0
torch-geometric==2.6.1
numba==0.60.0
```
Note: optimum-habana is essential, and torch should not be overwritten.
A new section in Dockerfile.hpu installs these packages, ensuring compatibility for HPU setup:

```Dockerfile
# Copy HPU requirements
COPY requirements_hpu.txt /root/whisper/requirements.txt

# Install Python packages
RUN pip install --upgrade pip \
    && pip install -r requirements.txt
```

We also added `.dockerignore` to avoid the problem: `ERROR: failed to solve: error from sender: open whisper/.graph_dumps: permission denied`
```
.graph_dumps
```

Please follow the README.md instructions under the section “Intel® Gaudi® HPU usage,” especially for:
Building the Docker Image: 
- Ensure the correct base image is used or updated to match your environment.
- Important: Before rebuilding, delete any existing Docker images with `docker image rm` to start fresh.

Running the Container:
- Set the volume path correctly (-v /path/to/your/whisper:/root/whisper).

Run with --device hpu for HPU compatibility in command-line usage:

Example:
```bash
python3 -m whisper.transcribe audi_file.wav --model turbo --device hpu
```

The results should be similar to:
```bash
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
============================= HABANA PT BRIDGE CONFIGURATION ===========================
 PT_HPU_LAZY_MODE = 1
 PT_RECIPE_CACHE_PATH =
 PT_CACHE_FOLDER_DELETE = 0
 PT_HPU_RECIPE_CACHE_CONFIG =
 PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
 PT_HPU_LAZY_ACC_PAR_MODE = 0
 PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
 PT_HPU_EAGER_PIPELINE_ENABLE = 1
 PT_HPU_EAGER_COLLECTIVE_PIPELINE_ENABLE = 1
---------------------------: System Configuration :---------------------------
Num CPU Cores : 96
CPU RAM       : 784291348 KB
------------------------------------------------------------------------------
/root/whisper/whisper/transcribe.py:132: UserWarning: FP16 is not supported on HPU; using FP32 instead
  warnings.warn("FP16 is not supported on HPU; using FP32 instead")
Detecting language using up to the first 30 seconds. Use `--language` to specify the language
Detected language: English
[00:00.800 --> 00:03.620]  Translated text...
```

Let me know if you have questions or run into issues during new setup!

orionsBeltWest commented:
Also in the command:

python3 -m whisper.transcribe audi_file.wav --model turbo --device hpu 


I do not see the audi_file.wav?  Is this messing or some other steps required?

orionsBeltWest commented:
Also I do not see the test dir so  I coul try to run the unittests ? I think this overwritten when mapped the /root/whisper to local whisper dir?

PiotrBLL committed:
Delete redundant linux packages in Dockerfile.hpu + update README.md

PiotrBLL committed:
Delete timestamps test

PiotrBLL committed:
Update Dockerfile.hpu and README.md files:
- rename `requirements_hpu.txt`,
- make docker run mapping optional,
- add running HPU tests in docs

PiotrBLL commented:
Hi @orionsBeltWest ,

We udated the Dockerfile.hpu (https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-8e8c9c8166fa53e2eb5a4fa3ea94131bf7e45e608f8161bbcee88c735fe171fb) following vllm (https://github.com/vllm-project/vllm/blob/main/Dockerfile.hpu) and also made an update in section Intel® Gaudi® hpu usage of README.md - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R147. Added an info about optional mapping using docker run command and also added an info about running the HPU tests inside the docker container - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R178

orionsBeltWest commented:
> Hi @orionsBeltWest ,
> 
> We udated the Dockerfile.hpu (https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-8e8c9c8166fa53e2eb5a4fa3ea94131bf7e45e608f8161bbcee88c735fe171fb) following vllm (https://github.com/vllm-project/vllm/blob/main/Dockerfile.hpu) and also made an update in section Intel® Gaudi® hpu usage of README.md - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R147. Added an info about optional mapping using docker run command and also added an info about running the HPU tests inside the docker container - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R178



No I am not able to run this build due to RuntimeError: synStatus=26 [Generic failure] Device acquire failed.

I ran it both in baremetal and k8 same issue. 


root@98ff692c3ac2:/workspace/whisper# python3 -m whisper.transcribe keynote.wav --model turbo --device hpu
/usr/lib/python3.10/runpy.py:126: RuntimeWarning: 'whisper.transcribe' found in sys.modules after import of package 'whisper', but prior to execution of 'whisper.transcribe'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
/usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
  return isinstance(object, types.FunctionType)
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
/workspace/whisper/whisper/__init__.py:154: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(fp, map_location="cpu")
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
/home/jenkins/workspace/cdsoftwarebuilder/create-binaries-from-sw-sources---bp-dt/repos/hcl/src/platform/gaudi_common/hcl_device_control_factory.cpp::84(initDevice): The condition [ g_ibv.init(deviceConfig) == hcclSuccess ] failed. ibv initialization failed 
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspace/whisper/whisper/transcribe.py", line 625, in <module>
    cli()
  File "/workspace/whisper/whisper/transcribe.py", line 597, in cli
    model = load_model(model_name, device=device, download_root=model_dir)
  File "/workspace/whisper/whisper/__init__.py", line 173, in load_model
    model = wrap_in_hpu_graph(model)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 720, in wrap_in_hpu_graph
    stream = htorch.hpu.default_stream()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/streams.py", line 284, in default_stream
    streamdata = _hpu_C._hpu_getDefaultStream(_get_device_index(device, optional=True))
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/_utils.py", line 42, in _get_device_index
    device_idx = hpu.current_device()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 132, in current_device
    init()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 71, in init
    _hpu_C.init()
RuntimeError: synStatus=26 [Generic failure] Device acquire failed.
Exception ignored in atexit callback: <built-in method cleanup of PyCapsule object at 0x7efb62607c30>
RuntimeError: Habana device not initialized
Exception ignored in atexit callback: <built-in method cleanup of PyCapsule object at 0x7efb62607c30>
RuntimeError: Habana device not initialized
root@98ff692c3ac2:/workspace/whisper# hl-smi 
+-----------------------------------------------------------------------------+
| HL-SMI Version:                              hl-1.18.0-fw-53.1.1.1          |
| Driver Version:                                     1.18.0-ee698fb          |
|-------------------------------+----------------------+----------------------+
| AIP  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | AIP-Util  Compute M. |
|===============================+======================+======================|
|   0  HL-225              N/A  | 0000:0e:00.0     N/A |                   0  |
| N/A   28C   N/A    72W / 600W |    768MiB / 98304MiB |     0%           N/A |
|-------------------------------+----------------------+----------------------+
| Compute Processes:                                               AIP Memory |
|  AIP       PID   Type   Process name                             Usage      |
|=============================================================================|
|   0        N/A   N/A    N/A                                      N/A        |
+=============================================================================+
root@98ff692c3ac2:/workspace/whisper#

orionsBeltWest commented:
I am able now to transcribe a wave file. 
 python3 -m whisper.transcribe keynote.wav --model turbo --device hpu


But when doing the pytest I am getting 14 failed test cases

======================================= short test summary info ========================================
FAILED tests/test_transcribe.py::test_transcribe_hpu[tiny.en] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[tiny] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[base.en] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[base] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[small.en] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[small] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[medium.en] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[medium] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v1] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v2] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v3] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[large] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v3-turbo] - KeyError: 'words'
FAILED tests/test_transcribe.py::test_transcribe_hpu[turbo] - KeyError: 'words'
============================= 14 failed, 29 warnings in 783.83s (0:13:03) ==============================

PiotrBLL committed:
Add word_timestamps fix in model.transcribe

PiotrBLL commented:
> I am able now to transcribe a wave file. python3 -m whisper.transcribe keynote.wav --model turbo --device hpu
> 
> But when doing the pytest I am getting 14 failed test cases
> 
> ======================================= short test summary info ======================================== FAILED tests/test_transcribe.py::test_transcribe_hpu[tiny.en] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[tiny] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[base.en] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[base] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[small.en] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[small] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[medium.en] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[medium] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v1] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v2] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v3] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[large] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[large-v3-turbo] - KeyError: 'words' FAILED tests/test_transcribe.py::test_transcribe_hpu[turbo] - KeyError: 'words' ============================= 14 failed, 29 warnings in 783.83s (0:13:03) ==============================

Made a fix of that, `word timestamps` were missing

orionsBeltWest reviewed (approved):
Tests passed. 
All good
Thanks.
